src/bp/migrations/v12_11_0-1592325280-migrate_say_nodes.ts

```diff
@@ -11,79 +13,131 @@ const migration: Migration = {
+    this.cmsService = this.inversify.get<CMSService>(TYPES.CMSService)
+    await this.cmsService.initialize()
+  }
+
+  async migrateBot(botId: string) {
+    const ghost = this.bp.ghost.forBot(botId)
+    const flows = await this.flowService.loadAll(botId)
+
+    for (const flow of flows) {
+      try {
+        for (const node of flow.nodes) {
+          this.migrateNode(node)
+        }
+
+        const flowContent = {
+          ..._.pick(flow, ['version', 'catchAll', 'startNode', 'skillData', 'triggers', 'label', 'description']),
+          nodes: flow.nodes.map(node => _.omit(node, 'x', 'y', 'lastModified'))
+        }
+        await ghost.upsertFile('./flows', flow.location!, JSON.stringify(flowContent, undefined, 2))
+      } catch (err) {
+        this.bp.logger
+          .forBot(botId)
+          .attachError(err)
+          .error('Could not migrate say node data')
+      }
+    }
+  }
+
+  migrateNode(node: sdk.FlowNode) {
+    if (node.type !== 'say_something' || node.contents || !node['content']) {
+      return
+    }
+
+    const { contentType, formData } = (node as any).content
+    const { newSchema } = this.cmsService.getContentType(contentType)
+    const schema = <sdk.FormDefinition>newSchema
+
+    const fields = [...schema.advancedSettings, ...schema.fields]
+    const root: any = { formData: { contentType } }
+
+    for (const [key, value] of Object.entries(formData)) {
+      const [name, lang] = key.split('$')
+      this.attachTranslation(root, 'formData', { [name]: value }, fields, lang)
+    }
+
+    delete (node as any).content
+    node.contents = [root.formData]
+
+    this.hasChanged = true
+  }
+
+  attachTranslation(root: any, basePath: string, current: any, fields: sdk.FormField[], lang: string) {
+    for (const [key, value] of Object.entries<any>(current)) {
+      const field = fields.find(x => x.key === key)
+      if (!field) {
+        continue
+      }
+
+      const path: string = field.translated ? `${basePath}.${key}.${lang}` : `${basePath}.${key}`
+
+      if (field.fields) {
+        const fields = this.getFieldsAndOptions(field)
+
+        if (field.type === 'group') {
+          for (let i = 0; i < value.length; i++) {
+            this.attachTranslation(root, `${path}.${i}`, value[i], fields, lang)
+          }
+        } else {
+          this.attachTranslation(root, path, value, fields, lang)
+        }
+      } else {
+        _.set(root, path, value)
+      }
+    }
+  }
+
+  getFieldsAndOptions(field: sdk.FormField): sdk.FormField[] {
+    if (!field.fields) {
+      return []
+    }
+    let fields = field.fields
+
+    for (const subfield of field.fields) {
+      if (subfield.options) {
+        for (const option of subfield.options) {
+          if (option.related) {
+            if (fields === field.fields) {
+              fields = [...field.fields]
+            }
+            fields.push(option.related)
+          }
+        }
+      }
+    }
```

Comment:
I don't understand what happens in this block of code. What does it do? 

Also, can you remove some level of indentation? And use another word than "field" which is litterally used multiple time on each lines

Reply:
The schema allows fields that declare options to also create other fields using the `related` option.

        {
            type: 'select',
            defaultValue: 'say',
            key: 'action',
            label: 'module.builtin.types.actionButton.actionLabel',
            options: [
              {
                value: 'say',
                label: 'module.builtin.types.actionButton.sayLabel',
                related: {
                  placeholder: 'module.builtin.types.actionButton.sayPlaceholder',
                  type: 'text',
                  key: 'text',
                  label: 'module.builtin.types.actionButton.sayTextLabel'
                }
              },
              {
                value: 'openUrl',
                label: 'module.builtin.types.actionButton.urlLabel',
                related: {
                  placeholder: 'module.builtin.types.actionButton.urlPlaceholder',
                  type: 'url',
                  key: 'text',
                  label: 'URL'
                }
              },
              {
                value: 'postBack',
                label: 'module.builtin.types.actionButton.postLabel',
                related: {
                  type: 'textarea',
                  key: 'text',
                  label: 'module.builtin.types.actionButton.postFieldLabel'
                }
              }
            ]
          }

So this function returns all fields defined by the schema + all the possible fields that could be created by each option.

It would be less confusing if those fields were declared like the other fields and just hidden when their respective options isn't selected.